### DISCUSSION TITLE
Reducing Boilerplate in Wagmi: Transitioning from Ethers and Typechain

### DISCUSSION THREAD

**Discussion by Jakkc** [wevm/wagmi#3968] [General]
I am transitioning to using wagmi after several years of working with ethers and typechain. I have noticed some patterns and potential challenges in this new paradigm and would appreciate any insights from the community.

In wagmi, it seems there is a significant amount of boilerplate overhead when using utilities, as each utility requires passing the config object. For example:

```
const block = await wagmiGetBlock(config);
const chainId = wagmiGetChainId(config);
const signedMessage = await wagmiSignMessage(config, { message });
```

This necessitates importing the config object each time a wagmi utility is used, which feels excessive compared to the abstraction provided in ethers.

To manage this, we have created hooks to wrap these utilities, passing the config object within the hook:

```
const useWeb3Client = () => {
  const config = useConfig();
  
  const getBlock = async () => {
    const block = await wagmiGetBlock(config);
    return block;
  };
}
```

While this approach works, it seems somewhat excessive to create custom hooks for basic actions like getBlock. Additionally, this pattern becomes more cumbersome for contract reads and writes:

```
const readContract = async ({ address, abi, functionName, args }) => {
  const result = await wagmiReadContract(config, { address, abi, functionName, args });
  return result;
};

const writeContract = async ({ address, abi, functionName, args, account }) => {
  const hash = await wagmiWriteContract(config, { address, abi, functionName, args, account });
  const txReceipt = await wagmiWaitForTransactionReceipt(config, { hash });
  return txReceipt;
};
```

Adopting this pattern results in losing the useful type inferences that wagmi provides. In comparison, typechain offers a straightforward API:

```
const contract = erc20.connect(address);
const foo = contract.doSomething(params);
```

In many cases with wagmi, it feels like we need to pass an excessive amount of boilerplate configuration for common actions.

I would like to know if there are best practices or alternative approaches to simplify this workflow in wagmi, or if this is the expected paradigm shift. Any guidance or insights from the community would be greatly appreciated as I work on building a new app with wagmi.

Thank you for your time and assistance.

**Comment by tmm** [comment]
Wagmi Core is optimized for the web so it uses tree-shakable actions so unused code can be eliminated from your final bundle, thus the need to explicitly pass Wagmi Config to Wagmi Core actions.

> To manage this, we have created hooks to wrap these utilities, passing the config object within the hook:

If you are already using hooks, I recommend checking out Wagmi's reactive primitives, either React Hooks or Vue Composables. For example, [`useReadContract`](https://wagmi.sh/react/api/hooks/useReadContract) handles dependency injecting the Wagmi Config via React Context so you don't need to do it manually.

> In comparison, typechain offers a straightforward API:

If you still want to use the builder pattern TypeChain and ethers use, you could try [`usePublicClient`](https://wagmi.sh/react/api/hooks/usePublicClient)/[`getPublicClient`](https://wagmi.sh/core/api/actions/getPublicClient) and [`useWalletClient`](https://wagmi.sh/react/api/hooks/useWalletClient)/[`getWalletClient`](https://wagmi.sh/core/api/actions/getWalletClient), or drop down to [Viem](https://viem.sh) entirely. Keep in mind that libraries and code that use the builder pattern, like TypeChain, ethers, and Viem's Public and Wallet Clients, can result in large bundle sizes (resulting in slow web load times for users).
